feat: Add `insert_link_hugr` to add entrypoint subtree AND link

acl-cqc opened the pull request:
Follows #2529

acl-cqc commented on hugr-core/src/hugr/linking.rs:
```diff
@@ -374,48 +475,185 @@ impl NameLinkingPolicy {
+        if use_entrypoint {
+            to_visit.push_back(source.entrypoint());
+            // Also add public defns not reachable from entrypoint that implement existing decls
+            // (ALAN this is a big API question. They might only be reachable from (the entrypoint via)
```
One option would be, add a `opportunistic_impls: bool` to the two `insert_link` methods; it's not really appropriate elsewhere. But this does make me wonder about doing reachability. If we have to traverse *both* Hugrs then we can do that but it is a bit of a PITA, and moreover, a pity 👎. There are an awful lot of variations one might consider, too 😞 😢 

I guess really the general case here is - if you want to add unreached functions, first you do `link_module` to add everything except the entrypoint; *then* you do `link_hugr` with various RaiseErrors and/or UseExistings  to make sure it doesn't add anything further.

Restricting to a subset of functions (by name, so you could do more-complex reachability yourself) is something I thought we might add later and was definitely hoping to leave that out of this PR. Just that this turned up in the first test I wrote....but maybe that's a bad case to have in a test. Thoughts???

ss2165 commented on hugr-core/src/hugr/linking.rs:
```diff
@@ -168,6 +168,99 @@ pub trait HugrLinking: HugrMut {
+    ///
+    /// If other's entrypoint calls a public [`FuncDefn`] in `other` which has the same name
+    /// and signature as a public [`FuncDefn`] in `self` and [`on_multi_defn`] is
+    /// [`OnMultiDefn::NewFunc`] of [`OnNewFunc::RaiseError`]
```
```suggestion
    /// [`OnMultiDefn::NewFunc`] or [`OnNewFunc::RaiseError`]
```

acl-cqc commented on hugr-core/src/hugr/linking.rs:
```diff
@@ -168,6 +168,99 @@ pub trait HugrLinking: HugrMut {
+    ///
+    /// # Errors
+    ///
+    /// If other's entrypoint is its module-root (recommend using [Self::link_module] instead)
```
No, it is an error...turned these into bullet points with explicit mention of the error variant raised

acl-cqc commented on hugr-core/src/hugr/linking.rs:
```diff
@@ -168,6 +168,99 @@ pub trait HugrLinking: HugrMut {
+    /// If other's entrypoint is its module-root (recommend using [Self::link_module] instead)
+    ///
+    /// If other's entrypoint calls (perhaps transitively) the function containing said entrypoint
+    /// (an exception is made if the called+containing function is public and is being replaced
```
Done...this para makes me think that perhaps `OnMultipleDefn::UseExisting` should be `::UseTarget` and `OnMultipleDefn::UseNew` should be `::UseSource`. Can do that rename is #2529, what do you think?

ss2165 replied:
Yes - pro consistency for identifying the two HUGRs either way

ss2165 reviewed (commented):
First review complete

acl-cqc commented on hugr-core/src/hugr/linking.rs:
```diff
@@ -338,82 +451,206 @@ impl NameLinkingPolicy {
+    fn action_for<SN: Display, TN: Copy + Display + std::fmt::Debug>(
+        &self,
+        existing: &HashMap<&str, PubFuncs<TN>>,
+        sn: SN,
```
Ok, now `existing in target` and `src_node`. There is some difficulty that we don't have a short name for a module-level child that gets copied by linking (so a func or a constant)...Node is too general, function is nearly there but doesn't cover constants.

acl-cqc commented:
Thanks @ss2165. Think I've addressed all but the one about the "most recent ancestor satisfying predicate"....

Any thoughts on factory methods? I've just removed `err_on_conflict` so we have `default` (errors on signature conflict or multiple defns)  and `new_keep_both_invalid` - a bit of a mouthful! 👎 - both allowing new names; plus mutators. (Also done in #2529)

Also, builder methods? I can include here, or another PR (there will be other follow-ups to use linking in more places!)

acl-cqc commented on hugr-core/src/builder.rs:
```diff
@@ -179,10 +179,16 @@ pub enum BuildError {
     },
 
     /// From [Dataflow::add_link_hugr_by_node_with_wires]
+    // TODO rename?
```
Deprecation done.
